### PR TITLE
Migrate auth to security context

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Added usage of security context if it passed in the request or launch spec which overrides the auth request or launch spec values.

Also modified the flyteadmin_config.yaml to use the right minio port and s3 bucket when running in local dev mode.
This makes it in sync with sandbox

Testing Done :
Created execution request from flytectl which passes in the security context.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/1722

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/workflowengine/impl/prepare_execution.go [hunk 1]
```diff
@@ -40,6 +40,21 @@ func addPermissions(auth *admin.AuthRole, roleNameKey string, flyteWf *v1alpha1.
 	}
 }
 
+func addPermissionsFromSecurtiyCtx(securityCtx *core.SecurityContext, roleNameKey string, flyteWf *v1alpha1.FlyteWorkflow) {
```
Comment (katrogan) [inline pkg/workflowengine/impl/prepare_execution.go:43]: what if you just pass the security context to addPermissions and have it prefer values from non-empty securitycontexts so that permissions handling is done in one place?

File: pkg/manager/impl/execution_manager.go [hunk 2]
```diff
@@ -672,6 +673,20 @@ func resolvePermissions(request *admin.ExecutionCreateRequest, launchPlan *admin
 	return &admin.AuthRole{}
 }
 
+func resolvePermissionsFromSecurityCtx(ctx context.Context, request admin.ExecutionCreateRequest, launchPlan *admin.LaunchPlan) *core.SecurityContext {
```
Comment (katrogan) [inline pkg/manager/impl/execution_manager.go:676]:
```suggestion
func resolveSecurityCtx(ctx context.Context, request admin.ExecutionCreateRequest, launchPlan *admin.LaunchPlan) *core.SecurityContext {
```

File: pkg/manager/impl/execution_manager.go [hunk 3]
```diff
@@ -672,6 +673,20 @@ func resolvePermissions(request *admin.ExecutionCreateRequest, launchPlan *admin
...
+	// Use security context from the request if its set
+	if request.Spec.SecurityContext != nil {
+		return request.Spec.SecurityContext
+	}
+
+	// Use launchplans security context if its set
+	if launchPlan.Spec.SecurityContext != nil {
+		return launchPlan.Spec.SecurityContext
```
Comment (katrogan) [inline pkg/manager/impl/execution_manager.go:684]: nit: you can always just return this value when `request.Spec.SecurityContext == nil`, no need to wrap it in an if block

File: flyteadmin_config.yaml [hunk 4]
```diff
@@ -64,7 +64,7 @@ database:
   port: 5432
   username: postgres
   host: localhost
-  dbname: postgres
+  dbname: flyteadmin
```
Comment (EngHabu) [inline flyteadmin_config.yaml:67]: Until we fix this: https://github.com/flyteorg/flyte/issues/1882, I don't think we should attempt to use anything but the default DB...

File: pkg/manager/impl/execution_manager.go [hunk 5]
```diff
@@ -672,6 +673,20 @@ func resolvePermissions(request *admin.ExecutionCreateRequest, launchPlan *admin
...
+	}
+	logger.Warn(ctx, "Returning an empty security context")
```
Comment (EngHabu) [inline pkg/manager/impl/execution_manager.go:686]: I think we should fall back on building Security Context from AuthRole if they are not specified in the request/launchplan... for the transition to deprecate AuthRole, I think Admin should be the place to make this translation...

File: flyteadmin_config.yaml (hunk 4 above)
Comment (pmahindrakar-oss) [reply flyteadmin_config.yaml]: Ok.reverted this to older values of postgress

File: pkg/manager/impl/execution_manager.go (hunk 5 above)
Comment (pmahindrakar-oss) [reply pkg/manager/impl/execution_manager.go]: Agreed. Done

File: pkg/manager/impl/execution_manager.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply pkg/manager/impl/execution_manager.go:687]: Actually there is additional logic due to which we cant return just the launplan.Spec.SecurityContext since we now want to get this value from resolvedAuthRole. Let me know if that works

```
// Use launchplans security context if its set
	if launchPlan.Spec.SecurityContext != nil {
		return launchPlan.Spec.SecurityContext
	}
	logger.Warn(ctx, "Setting security context from auth Role")
	return &core.SecurityContext{
		RunAs: &core.Identity{
			IamRole:           resolvedAuthRole.AssumableIamRole,
			K8SServiceAccount: resolvedAuthRole.KubernetesServiceAccount,
		},
	}
```

File: pkg/manager/impl/execution_manager.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply pkg/manager/impl/execution_manager.go]: Done

File: pkg/workflowengine/impl/prepare_execution.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply pkg/workflowengine/impl/prepare_execution.go:49]: Done .Added a wrapper addPermissions which accepts both authRole and securityCtx

File: pkg/manager/impl/execution_manager.go (hunk 3 above)
Comment (katrogan) [reply pkg/manager/impl/execution_manager.go:687]: looks great!

**Comment by pmahindrakar-oss** [comment]
PTAL @katrogan @EngHabu

File: pkg/manager/impl/execution_manager.go [hunk 6]
```diff
@@ -568,13 +568,16 @@ func (m *ExecutionManager) launchSingleTaskExecution(
...
+	resolvedAuthRole := resolvePermissions(request, launchPlan)
+	resolvedSecurityCtx := resolveSecurityCtx(ctx, request, launchPlan, resolvedAuthRole)
 	executionParameters := workflowengineInterfaces.ExecutionParameters{
 		Inputs:              request.Inputs,
 		AcceptedAt:          requestedAt,
 		Labels:              labels,
 		Annotations:         annotations,
 		ExecutionConfig:     executionConfig,
-		Auth:                resolvePermissions(&request, launchPlan),
+		Auth:                resolvedAuthRole,
```
Comment (EngHabu) [inline pkg/manager/impl/execution_manager.go:579]: So is it possible to just remove Auth field completely from ExecutionParameters? is this used anywhere that's not FlyteAdmin? if it only lives in flyteAdmin, then we can update the code to always use the SecurityContext and we make sure it's populated here.... right?

Comment (katrogan) [reply pkg/manager/impl/execution_manager.go:579]: +1 we should move away from passing Auth objects

Comment (pmahindrakar-oss) [reply pkg/manager/impl/execution_manager.go]: Make sense. Its not used outside admin. Have removed it from ExecutionParameters. It will remain in the ExecutionCreateRequest though until all clients migrate after which we can remove it completely along with the protos.

Comment (pmahindrakar-oss) [reply pkg/manager/impl/execution_manager.go]: Done

Comment (katrogan) [reply pkg/manager/impl/execution_manager.go]: one more thing to remove from flyteidl in flyte 1.0.0 😄

**Review by katrogan — APPROVED** [review]
